# 4.8 Grouping data and aggregating variables - Part 1

## This script contains the following points:

### Step 1 - Import data
### Step 2 - Group data
### Step 3 - Analyze results
### Step 4 - Create a customer loyalty flag
### Step 5 - Create basic statistics
### Step 6 - Create a spending flag
### Step 7 - Create an order frequency flag
### Step 9 - Export data

In [1]:
# Step 1 - Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Assign project folder path to a variable

path = r'C:\Users\jomok\Documents\Career Foundry\Achievement 4\07-2023 Instacart Basket Analysis'

## Step 1. Import data

In [3]:
# Import data

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# Check output

ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... prices    price_range_loc     busiest_day  \
0          0  ...    9.0  Mid-range product  Regularly busy   
1          1  ...    9.0  Mid-range product    Slowest days   
2          1  ...    9.0  Mid-range product    Slowest days   
3          1  ...    9.0  Mid-range product    Slowest days   
4          1  ...    9.0  Mid-range product    Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_price  spending_flag  \
0         Average orders        10  New customer  6.367797    Low spender   
1         Average orders        10  New customer  6.367797    Low spender   
2            Most orders        10  New customer  6.367797    Low spender   
3         Average orders        10  New customer  6.367797    Low spender   
4            Most orders        10  New customer  6.367797    Low spender   

  median_days_since_prior_order  order_frequency_flag  
0                          20.0      Regular customer  
1                          20.0      Regular customer  
2                          20.0      Regular customer  
3                          20.0      Regular customer  
4                          20.0      Regular customer  

[5 rows x 23 columns]

In [5]:
# Check shape

ords_prods_merge.shape

(32404859, 23)

## Step 2 - Group data

In [6]:
# Split data into "department_id" groups and aggregate mean values by "order_number" column

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3 - Results for the entire dataframe contains output for every department, whereas the first million rows was not representative of all departments.  The mean value range also increased to now 15.22 to 22.90 for the entire dataframe, as opposed to a smaller range of 17.47 to 19.46 for the 1 million first rows subset dataframe.

## Step 4 - Create a customer loyalty flag

In [7]:
# Create "max_order" column

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# Check output

ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered  ... prices    price_range_loc     busiest_day  \
0           0  ...    9.0  Mid-range product  Regularly busy   
1           1  ...    9.0  Mid-range product    Slowest days   
2           1  ...    9.0  Mid-range product    Slowest days   
3           1  ...    9.0  Mid-range product    Slowest days   
4           1  ...    9.0  Mid-range product    Slowest days   
5           1  ...    9.0  Mid-range product  Regularly busy   
6           1  ...    9.0  Mid-range product    Busiest days   
7           1  ...    9.0  Mid-range product    Busiest days   
8           1  ...    9.0  Mid-range product    Busiest days   
9           1  ...    9.0  Mid-range product    Slowest days   
10          0  ...    9.0  Mid-range product    Busiest days   
11          1  ...    9.0  Mid-range product  Regularly busy   
12          1  ...    9.0  Mid-range product  Regularly busy   
13          1  ...    9.0  Mid-range product    Busiest days   
14          1  ...    9.0  Mid-range product    Busiest days   

    busiest_period_of_day max_order      loyalty_flag avg_price  \
0          Average orders        10      New customer  6.367797   
1          Average orders        10      New customer  6.367797   
2             Most orders        10      New customer  6.367797   
3          Average orders        10      New customer  6.367797   
4             Most orders        10      New customer  6.367797   
5          Average orders        10      New customer  6.367797   
6             Most orders        10      New customer  6.367797   
7             Most orders        10      New customer  6.367797   
8             Most ord

In [9]:
# Tell pandas not to assign any options for maximum number of rows to display

pd.options.display.max_rows = None

In [10]:
# Check output for 100 rows

ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   
15   2293453       19    prior             2                  5   
16   1973799       19    prior             5                  6   
17    532817       19    prior             7                  4   
18   1573906       21    prior            10                  3   
19   1593000       31    prior            10                  3   
20   2231262       31    prior            17                  3   
21   2580647       43    prior             6                  4   
22   2187180       43    prior             9                  4   
23   2497897       52    prior             1                  1   
24   1318871       52    prior             2                  1   
25   1261384       52    prior             3                  1   
26    580568       52    prior             5                  2   
27   2428073       52    prior             7                  2   
28    342306       52    prior             9                  1   
29   1520353       52    prior            10                  2   
30   2542086       52    prior            11                  2   
31    944694       52    prior            12                  1   
32   1268191       52    prior            15                  2   
33   1122089       52    prior            17                  2   
34   1498922       52    prior            18                  2   
35    180919       52    prior            19                  3   
36   2695875       52    prior            21                  5   
37   3205960       67    prior             1                  3   
38    970922       67    prior             2                  1   
39    596585       67    prior             3                  1   
40   2037087       67    prior             4                  2   
41   2474111       67    prior             5                  3   
42   1568481       67    prior             6                  3   
43   2048527       67    prior             8                  2   
44   2389031       67    prior             9                  2   
45   2378562       67    prior            10                  1   
46   1515698       67    prior            11                  1   
47   1928050       67    prior            12                  2   
48   1949274       67    prior            13                  1   
49    551869       67    prior            14                  1   
50    768572       67    prior            16                  4   
51    150528       67    prior            18                  5   
52     29758       67    prior            19                  0   
53   1558976       67    prior            21                  1   
54    863974       67    prior            22                  1   
55   2308399       67    prior            23                  1   
56   2759128       81    prior             3                  1   
57   1812985       81    prior             5                  2   
58   1889835       82    prior             1       

In [11]:
# Create flag to assign "loyalty" label

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# Check frequency of "loyalty_flag" column

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [15]:
# Check output for first 60 rows of "user_id", "loyalty_flag" and "order_number" columns

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

## Step 5 - Check basic statistics

In [16]:
# Check statistics of "loyalty_flag" column

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

The prices of products purchased by loyal customers are lower on average than those purchased by regular or new customers.

## Step 6 - Create a spending flag

In [17]:
# Group data

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [18]:
# Check output

ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered  ... prices    price_range_loc     busiest_day  \
0           0  ...    9.0  Mid-range product  Regularly busy   
1           1  ...    9.0  Mid-range product    Slowest days   
2           1  ...    9.0  Mid-range product    Slowest days   
3           1  ...    9.0  Mid-range product    Slowest days   
4           1  ...    9.0  Mid-range product    Slowest days   
5           1  ...    9.0  Mid-range product  Regularly busy   
6           1  ...    9.0  Mid-range product    Busiest days   
7           1  ...    9.0  Mid-range product    Busiest days   
8           1  ...    9.0  Mid-range product    Busiest days   
9           1  ...    9.0  Mid-range product    Slowest days   
10          0  ...    9.0  Mid-range product    Busiest days   
11          1  ...    9.0  Mid-range product  Regularly busy   
12          1  ...    9.0  Mid-range product  Regularly busy   
13          1  ...    9.0  Mid-range product    Busiest days   
14          1  ...    9.0  Mid-range product    Busiest days   

    busiest_period_of_day max_order      loyalty_flag avg_price  \
0          Average orders        10      New customer  6.367797   
1          Average orders        10      New customer  6.367797   
2             Most orders        10      New customer  6.367797   
3          Average orders        10      New customer  6.367797   
4             Most orders        10      New customer  6.367797   
5          Average orders        10      New customer  6.367797   
6             Most orders        10      New customer  6.367797   
7             Most orders        10      New customer  6.367797   
8             Most ord

In [19]:
# Create flag to assign "spending" label

ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [20]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [21]:
# Check frequency of "spending_flag" column

ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [22]:
# Check output for first 60 rows of "user_id", "spending_flag" and "order_number" columns

ords_prods_merge[['user_id', 'spending_flag', 'order_number']].head(60)

user_id spending_flag  order_number
0         1   Low spender             1
1         1   Low spender             2
2         1   Low spender             3
3         1   Low spender             4
4         1   Low spender             5
5         1   Low spender             6
6         1   Low spender             7
7         1   Low spender             8
8         1   Low spender             9
9         1   Low spender            10
10       15   Low spender            15
11       15   Low spender            17
12       15   Low spender            18
13       15   Low spender            21
14       15   Low spender            22
15       19   Low spender             2
16       19   Low spender             5
17       19   Low spender             7
18       21   Low spender            10
19       31   Low spender            10
20       31   Low spender            17
21       43   Low spender             6
22       43   Low spender             9
23       52   Low spender             1
24       52   Low spender             2
25       52   Low spender             3
26       52   Low spender             5
27       52   Low spender             7
28       52   Low spender             9
29       52   Low spender            10
30       52   Low spender            11
31       52   Low spender            12
32       52   Low spender            15
33       52   Low spender            17
34       52   Low spender            18
35       52   Low spender            19
36       52   Low spender            21
37       67   Low spender             1
38       67   Low spender             2
39       67   Low spender             3
40       67   Low spender             4
41       67   Low spender             5
42       67   Low spender             6
43       67   Low spender             8
44       67   Low spender             9
45       67   Low spender            10
46       67   Low spender            11
47       67   Low spender            12
48       67   Low spender            13
49       67   Low spender            14
50       67   Low spender            16
51       67   Low spender            18
52       67   Low spender            19
53       67   Low spender            21
54       67   Low spender            22
55       67   Low spender            23
56       81   Low spender             3
57       81   Low spender             5
58       82   Low spender             1
59       82   Low spender             3

## Step 7 - Create an order frequency flag

In [23]:
# Group data

ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [24]:
# Check output

ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     0.0         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered  ... prices    price_range_loc     busiest_day  \
0           0  ...    9.0  Mid-range product  Regularly busy   
1           1  ...    9.0  Mid-range product    Slowest days   
2           1  ...    9.0  Mid-range product    Slowest days   
3           1  ...    9.0  Mid-range product    Slowest days   
4           1  ...    9.0  Mid-range product    Slowest days   
5           1  ...    9.0  Mid-range product  Regularly busy   
6           1  ...    9.0  Mid-range product    Busiest days   
7           1  ...    9.0  Mid-range product    Busiest days   
8           1  ...    9.0  Mid-range product    Busiest days   
9           1  ...    9.0  Mid-range product    Slowest days   
10          0  ...    9.0  Mid-range product    Busiest days   
11          1  ...    9.0  Mid-range product  Regularly busy   
12          1  ...    9.0  Mid-range product  Regularly busy   
13          1  ...    9.0  Mid-range product    Busiest days   
14          1  ...    9.0  Mid-range product    Busiest days   

    busiest_period_of_day max_order      loyalty_flag avg_price  \
0          Average orders        10      New customer  6.367797   
1          Average orders        10      New customer  6.367797   
2             Most orders        10      New customer  6.367797   
3          Average orders        10      New customer  6.367797   
4             Most orders        10      New customer  6.367797   
5          Average orders        10      New customer  6.367797   
6             Most orders        10      New customer  6.367797   
7             Most orders        10      New customer  6.367797   
8             Most ord

In [25]:
# Create flag to assign "order_frequency" label

ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [26]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [27]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [28]:
# Check frequency of "order_frequency_flag" column

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: order_frequency_flag, dtype: int64

In [29]:
# Check output for first 60 rows of "user_id", "order_frequency_flag" and "order_number" columns

ords_prods_merge[['user_id', 'order_frequency_flag', 'order_number']].head(60)

user_id   order_frequency_flag  order_number
0         1       Regular customer             1
1         1       Regular customer             2
2         1       Regular customer             3
3         1       Regular customer             4
4         1       Regular customer             5
5         1       Regular customer             6
6         1       Regular customer             7
7         1       Regular customer             8
8         1       Regular customer             9
9         1       Regular customer            10
10       15      Frequent customer            15
11       15      Frequent customer            17
12       15      Frequent customer            18
13       15      Frequent customer            21
14       15      Frequent customer            22
15       19      Frequent customer             2
16       19      Frequent customer             5
17       19      Frequent customer             7
18       21      Frequent customer            10
19       31      Frequent customer            10
20       31      Frequent customer            17
21       43      Frequent customer             6
22       43      Frequent customer             9
23       52      Frequent customer             1
24       52      Frequent customer             2
25       52      Frequent customer             3
26       52      Frequent customer             5
27       52      Frequent customer             7
28       52      Frequent customer             9
29       52      Frequent customer            10
30       52      Frequent customer            11
31       52      Frequent customer            12
32       52      Frequent customer            15
33       52      Frequent customer            17
34       52      Frequent customer            18
35       52      Frequent customer            19
36       52      Frequent customer            21
37       67      Frequent customer             1
38       67      Frequent customer             2
39       67      Frequent customer             3
40       67      Frequent customer             4
41       67      Frequent customer             5
42       67      Frequent customer             6
43       67      Frequent customer             8
44       67      Frequent customer             9
45       67      Frequent customer            10
46       67      Frequent customer            11
47       67      Frequent customer            12
48       67      Frequent customer            13
49       67      Frequent customer            14
50       67      Frequent customer            16
51       67      Frequent customer            18
52       67      Frequent customer            19
53       67      Frequent customer            21
54       67      Frequent customer            22
55       67      Frequent customer            23
56       81  Non-frequent customer             3
57       81  Non-frequent customer             5
58       82       Regular customer             1
59       82       Regular customer             3

## Step 9 - Export data

In [30]:
# Export dataframe to pkl

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))